servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE DeriveDataTypeable  #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+-- | A module providing regexp validated textual data.
+--
+-- This module is intended to imported qualified
+--
+-- > import qualified Servant.Contrib.RegexHttpApiData as RE
+--
+-- The 'RE' type can be used, for example to do quick'n'dirty @servant@ apis.
+-- In the long run it's valuable to write specific types for 'Capture's or
+-- 'QueryParam's, but this type can help at the beginning.
+--
+-- @
+-- type API = "data" :> Capture "city" (RE.RE "a-zA-Z+") :> Get '[JSON] SomeData
+-- @
+module Servant.Contrib.RegexHttpApiData (
+    RE,
+    mk,
+    original,
+    groups,
+    ) where
+
+import           Data.Maybe
+                 (fromJust)
+import           Data.Proxy
+                 (Proxy (..))
+import qualified Data.Text       as T
+import           Data.Typeable
+                 (Typeable)
+import           GHC.TypeLits
+                 (KnownSymbol, Symbol, symbolVal)
+import qualified Text.Regex.TDFA as R
+import           Web.HttpApiData
+                 (FromHttpApiData (..), ToHttpApiData (..))
+
+-- | A 'Text' value "smartly" constructed to match the given reg-exp.
+--
+-- It also contains a lazily construted list of capture groups.
+--
+-- === Examples
+--
+-- >>> mk "foo" :: Maybe (RE "^foo$")
+-- Just (RE "foo" [["foo"]])
+--
+-- >>> mk "abc" :: Maybe (RE "(abc")
+-- Nothing
+--
+-- >>> mk "abc" :: Maybe (RE "foo")
+-- Nothing
+data RE (re :: Symbol) = RE !T.Text [[T.Text]]
+  deriving (Show, Typeable)
+
+-- | Construct 'RE' value.
+-- If regexp is invalid, or value doesn't match it 'Nothing' is returned.
+mk :: forall re. KnownSymbol re => T.Text -> Maybe (RE re)
```

Comment:
Can we distinguish "invalid regexp" and "value doesn't match it" cases?
Can we detect "invalid regexp" at compile time?